lib/enum.ml

```diff
@@ -0,0 +1,204 @@
+    let cases = List.map (from_string_case_from_constructor ~loc) constructors in
+    let cases = cases @ [invalid_case_for_from_string ~loc ~function_name] in
+    let expr = Ast_builder.Default.pexp_function ~loc cases in
+    let value_description =
+      Ast_builder.Default.value_binding
+      ~loc
+      ~pat
+      ~expr
+    in
+    Ast_builder.Default.pstr_value ~loc Nonrecursive [value_description]
+
+  let from_enummable_variant
+    ~loc
+    ~type_name
+    ~constructors
+  =
+    [ to_string_function ~loc ~type_name ~constructors
+    ; from_string_function ~loc ~type_name ~constructors
+    ]
+
+  let from_type_declaration ~loc type_ =
+    match type_ with
+    | {ptype_kind = Ptype_record _; _}
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "record"
+    | {ptype_kind = Ptype_abstract; _}
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "abstract"
+    | {ptype_kind = Ptype_open; _}
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "open"
+    | {ptype_kind = Ptype_variant _; ptype_params = _::_; _}
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "parametrized variant"
+    | {ptype_kind = Ptype_variant constructors; _} when not (Utils.constructors_are_bare constructors)
+      ->
+        Raise.Enum.unhandled_type_kind ~loc "variant with arguments"
+    | {ptype_kind = Ptype_variant constructors; ptype_name = {txt = type_name; _}; _}
+      ->
+        from_enummable_variant ~loc ~type_name ~constructors
+
+  (** By giving this to the Deriving.Generator.make_noarg function below, ppxlib
+   *  will apply the function the parameters:
+   *  ~loc: Information about the current location in the code base (file, lineno etc)
+   *  ~path: The current file path?
+   *  rec_flag: ???
+   *  type_declarations: A list of the type declarations at the point of call
+   *)
+  let from_type_decl ~loc ~path:_ (_rec_flag, type_declarations) =
+    List.flatten @@ List.map (from_type_declaration ~loc) type_declarations
+end
+
+module Sig = struct
+  let to_string_function_val ~loc ~type_name =
+    let function_name = Utils.to_string_function_name ~enum_name:type_name in
+    let type_lident = {txt = Lident type_name; loc} in
+    let lhs_type = Ast_builder.Default.ptyp_constr ~loc type_lident [] in
+    let type_ =
+      Ast_builder.Default.ptyp_arrow
+        ~loc
+        Nolabel
+        lhs_type
+        [%type: string]
+    in
+    let value_description =
+      Ast_builder.Default.value_description
+      ~loc
+      ~name:{txt = function_name; loc}
+      ~type_
+      ~prim:[]
+    in
+    Ast_builder.Default.psig_value ~loc value_description
+
+  let from_string_function_val ~loc ~type_name =
+    let function_name = Utils.from_string_function_name ~enum_name:type_name in
+    let type_lident = {txt = Lident type_name; loc} in
+    let rhs_type = Ast_builder.Default.ptyp_constr ~loc type_lident [] in
+    let type_ =
+      Ast_builder.Default.ptyp_arrow
```

Comment:
Same here, you can probably rewrite that as:
```ocaml
let type_ = [%type: string -> [%t rhs_type]]
```
alternatively you could maybe even go further and return:
```ocaml
[%sigi: val from_string : string -> [%type rhs_type]]
``` 
so you don't need to build `type_` and `value_description`.
Not sure about the exact syntax for `[%sigi: ...]` but we can look into together if you want!

Reply:
Same thing apply for the `to_string` version as well.